SYCL/Matrix/element_wise_add_bf16_jm_apply_impl.hpp

```diff
@@ -0,0 +1,133 @@
+
+#define TM 8
+#define TN SG_SZ
+#define TK 16
+
+static float make_fp32(bfloat16 x) {
+  unsigned int y = *((int *)&x);
+  y = y << 16;
+  float *res = reinterpret_cast<float *>(&y);
+  return *res;
+}
+
+template <typename T, size_t NUM_ROWS, size_t NUM_COLS> struct big_matrix {
+public:
+  T *mat;
+
+public:
+  T *get_data() { return mat; }
+  void set_data(T *data) { mat = data; }
+  big_matrix(T *data) : mat(data) {}
+};
+
+template <typename T, size_t M, size_t N>
+void assert_ops_ref(
+    accessor<T, 2, access::mode::read, access::target::host_buffer> C,
+    const float ref) {
+  for (size_t i = 0; i < M; i++)
+    for (size_t j = 0; j < N; j++) {
+      auto diff = make_fp32(C[i][j]) - ref;
+      assert(std::fabs(static_cast<float>(diff)) <
+             std::numeric_limits<float>::epsilon());
+    }
+}
+
+template <typename T>
+struct apply_add {
+  void operator() (T &x) {
+    x = x + bfloat16(2);
+  }
+};
+
+template <typename T, size_t M, size_t N>
+void matrix_verify_add_lambda(queue q, big_matrix<T, M, N> &A, nd_range<2> &r,
+                       const float ref) {
+  buffer<bfloat16, 2> bufA(A.get_data(), range<2>(M, N));
+
+  q.submit([&](handler &cgh) {
+     auto accA = bufA.get_access<access::mode::read_write>(cgh);
+
+     cgh.parallel_for<class add_matrix_lambda>(
+         r, [accA](nd_item<2> spmd_item) [[intel::reqd_sub_group_size(SG_SZ)]] {
+           const auto global_idx = spmd_item.get_global_id(0);
+           const auto global_idy = spmd_item.get_global_id(1);
+           const auto sg_startx = global_idx - spmd_item.get_local_id(0);
+           const auto sg_starty = global_idy - spmd_item.get_local_id(1);
+
+           sub_group sg = spmd_item.get_sub_group();
+           joint_matrix<sub_group, T, use::a, TM, TK, layout::row_major> sub_a;
+
+           joint_matrix_fill(sg, sub_a, bfloat16(5.0));
+
+           joint_matrix_apply(sg, sub_a, [=](T &x){x = x + bfloat16(2);});
```

Comment:
Can we make that lambda a param of the `matrix_verify` to avoid code duplication between `matrix_verify_add_lambda`/`matrix_verify_add_func_obj`?

I also can't imagine why the behavior could be different at all. After all, lambda is just a syntactic sugar for the creation of this function object. I think a function pointer would make more sense for the extra test...

Reply:
@aelovikov-intel , thanks for the suggestion and your suggestion makes sense.  I tried to have one single `matrix_verify()`, which accepts a `F&&` as a parameter.  Then I tried calling it twice, one with a lambda and one with a Function object:

`matrix_verify_add<bfloat16, MATRIX_M, MATRIX_N>(q, MA, r, 7.0, [=](bfloat16 &x) { x = x + bfloat16(2); });`
`matrix_verify_add<bfloat16, MATRIX_M, MATRIX_N>(q, MA, r, 7.0, apply_add<bfloat16>());`

This gives rise to compilation error such as:

```
/tmp/joint_matrix_apply_bf16-header-a0f980.h:87:20: error: redefinition of 'KernelInfo<add_matrix>'
template <> struct KernelInfo<add_matrix> {
                   ^~~~~~~~~~~~~~~~~~~~~~
/tmp/joint_matrix_apply_bf16-header-a0f980.h:40:20: note: previous definition is here
template <> struct KernelInfo<add_matrix> {
```

Any suggestion on how to break this?

Reply:
Regarding the function pointer test case, we are not supporting function pointer for `joint_matrix_apply()`.  We only support lambdas and function objects.

Reply:
> @aelovikov-intel , thanks for the suggestion and your suggestion makes sense. I tried to have one single `matrix_verify()`, which accepts a `F&&` as a parameter. Then I tried calling it twice, one with a lambda and one with a Function object:
> 
> `matrix_verify_add<bfloat16, MATRIX_M, MATRIX_N>(q, MA, r, 7.0, [=](bfloat16 &x) { x = x + bfloat16(2); });` `matrix_verify_add<bfloat16, MATRIX_M, MATRIX_N>(q, MA, r, 7.0, apply_add<bfloat16>());`
> 
> This gives rise to compilation error such as:
> 
> ```
> /tmp/joint_matrix_apply_bf16-header-a0f980.h:87:20: error: redefinition of 'KernelInfo<add_matrix>'
> template <> struct KernelInfo<add_matrix> {
>                    ^~~~~~~~~~~~~~~~~~~~~~
> /tmp/joint_matrix_apply_bf16-header-a0f980.h:40:20: note: previous definition is here
> template <> struct KernelInfo<add_matrix> {
> ```
> 
> Any suggestion on how to break this?

Nevermind, Dounia suggested a solution of not using any name and it worked.